### PR TITLE
Static analysis upgrade testing

### PR DESCRIPTION
**Pull request by dkjellin**
Using static analysis for upgrade testing. Adding two new annotations to handle how we process upgrades,

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -238,47 +240,7 @@
             <version>1.2</version>
             <scope>runtime</scope>
         </dependency>
-
-        <dependency>
-            <groupId>org.hamcrest</groupId>
```
Comment (dkjellin) [inline pom.xml:243]: I ordered dependencies alphabetically to make it easier to find

File: pom.xml [hunk 2]
```diff
@@ -312,6 +274,58 @@
...
+                    <artifactId>commons-lang3</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>com.sun.xml.bind</groupId>
+                    <artifactId>jaxb-osgi</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>net.lingala.zip4j</groupId>
```
Comment (dkjellin) [inline pom.xml:300]: this is a new dependency

File: pom.xml [hunk 3]
```diff
@@ -312,6 +274,58 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>2.21.0</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.reflections</groupId>
```
Comment (dkjellin) [inline pom.xml:324]: this is a new dependency

File: pom.xml [hunk 4]
```diff
@@ -533,6 +550,22 @@
                 </plugins>
             </build>
         </profile>
+        <profile>
+            <id>upgradeTest</id>
```
Comment (dkjellin) [inline pom.xml:554]: new profile so it doesn't run unless you want it to

**Comment by dkjellin** [comment]
@mhenschke-atl We spoke of adding annotations to o-auth classes as they are not upgrade sensitive. I have not done that, what do you think shall I do it in this PR, or in a smaller targeted one?

File: pom.xml [hunk 5]
```diff
@@ -333,6 +347,7 @@
                          exclude integration tests here. -->
                     <excludes>
                         <exclude>it/**/*.java</exclude>
+                        <exclude>upgrade/**/*.java</exclude>
```
Comment (dkjellin) [inline pom.xml:350]: this exclusion can now be removed, as it is part of the property on the line below, I'll update

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/annotations/NotUpgradeSensitive.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import static java.lang.annotation.ElementType.TYPE;
+
+/**
+ * Indicates that the class in question is not subject to upgrade concerns. This means it is not part of a job in any
+ * shape or form. It is not persisted in any way. Examples include Actions that displays a help link.
+ * @since 3.0.0
+ */
+@Target({TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface NotUpgradeSensitive {
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/annotations/NotUpgradeSensitive.java:16]: I noticed this isn't checked in the `testAnnotationsAreAllStillRelevant`. Do we intend for this annotation to persist, even after the plugin is released?

File: src/test/java/upgrade/com/atlassian/bitbucket/jenkins/internal/SafeRemovedClassList.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package upgrade.com.atlassian.bitbucket.jenkins.internal;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SafeRemovedClassList {
```
Comment (mhenschke-atl) [inline src/test/java/upgrade/com/atlassian/bitbucket/jenkins/internal/SafeRemovedClassList.java:9]: Do I understand correctly that the safe removed class list is for classes that we've authored, have changes that break the test but we are confident will not impact the upgrade flow?

**Review by mhenschke-atl — APPROVED** [review]
This looks good, I just had a few questions.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/annotations/NotUpgradeSensitive.java (hunk 6 above)
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/annotations/NotUpgradeSensitive.java:16]: Yes, I didn't really add many in this PR, but things like the `OAuthConsumerCreateAction` is not upgrade sensitive as it is not persisted. So these are marked "once and for all". We exclude them in the initial scan, but if they are included in a persisted class they are still checked.

File: src/test/java/upgrade/com/atlassian/bitbucket/jenkins/internal/SafeRemovedClassList.java (hunk 7 above)
Comment (dkjellin) [reply src/test/java/upgrade/com/atlassian/bitbucket/jenkins/internal/SafeRemovedClassList.java:9]: Pretty much, it is for changes that cannot be annotated (say a removal of a class) that you've manually tested and deemed safe for upgrade.

File: src/test/java/upgrade/com/atlassian/bitbucket/jenkins/internal/FailureCollector.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    public void addAnnotationRemovalFailure(String version, Field field) {
+        errors.add(format("UpgradeHandled annotaton on '%s' in '%s' should be removed in or before current version (%s)",
+                field.getName(), field.getDeclaringClass().getName(), version));
+    }
+
+    public void addClassNotFound(ClassNotFoundException e) {
+        errors.add(format("Could not load '%s' this will cause an upgrade error", e.getMessage()));
+    }
+
+    public void addFieldDifference(Field field) {
```
Comment (ysun-atl) [inline src/test/java/upgrade/com/atlassian/bitbucket/jenkins/internal/FailureCollector.java:33]: This method is not used

**Review by ysun-atl — APPROVED** [review]

**Pull request merged**